ethereum/blockwatch/block_watcher.go

```diff
@@ -185,46 +194,82 @@ func (w *Watcher) GetAllRetainedBlocks() ([]*miniheader.MiniHeader, error) {
 	return w.stack.PeekAll()
 }
 
-// pollNextBlock polls for the next block header to be added to the block stack.
-// If there are no blocks on the stack, it fetches the first block at the specified
-// `startBlockDepth` supplied at instantiation.
-func (w *Watcher) pollNextBlock() error {
-	var nextBlockNumber *big.Int
-	latestHeader, err := w.stack.Peek()
+// syncChain syncs our local state of the chain to the latest block found via
+// Ethereum RPC
+func (w *Watcher) syncChain() error {
+	latestHeader, err := w.client.HeaderByNumber(nil)
 	if err != nil {
 		return err
 	}
-	if latestHeader == nil {
-		if w.startBlockDepth == rpc.LatestBlockNumber {
-			nextBlockNumber = nil // Fetch latest block
-		} else {
-			nextBlockNumber = big.NewInt(int64(w.startBlockDepth))
-		}
-	} else {
-		nextBlockNumber = big.NewInt(0).Add(latestHeader.Number, big.NewInt(1))
-	}
-	nextHeader, err := w.client.HeaderByNumber(nextBlockNumber)
+	lastStoredHeader, err := w.stack.Peek()
 	if err != nil {
-		if err == ethereum.NotFound {
-			log.WithFields(log.Fields{
-				"blockNumber": nextBlockNumber,
-			}).Trace("block header not found")
-			return nil // Noop and wait next polling interval
-		}
 		return err
 	}
 
-	events := []*Event{}
-	events, err = w.buildCanonicalChain(nextHeader, events)
-	// Even if an error occurred, we still want to emit the events gathered since we might have
-	// popped blocks off the Stack and they won't be re-added
-	if len(events) != 0 {
-		w.blockFeed.Send(events)
+	var numBlocksToFetch int
+	// No blocks stored yet, fetch the first
+	if lastStoredHeader == nil {
+		numBlocksToFetch = 1
+	} else {
+		// Noop if already caught up or ahead of latest block returned from Ethereum node
+		if latestHeader.Number.Int64() <= lastStoredHeader.Number.Int64() {
+			return nil
+		}
+		numBlocksToFetch = int(latestHeader.Number.Int64() - lastStoredHeader.Number.Int64())
 	}
-	if err != nil {
-		return err
+
+	allEvents := []*Event{}
+	var syncErr error
+	for i := 0; i < numBlocksToFetch; i++ {
```

Comment:
Added the following conditional:

```
if numBlocksToFetch >= constants.MaxBlocksStoredInNonArchiveNode {
		return fmt.Errorf(
			"cannot sync blocks more than %d blocks behind latest (%d). Your `BLOCK_POLLING_INTERVAL` is probably set too high",
			constants.MaxBlocksStoredInNonArchiveNode,
			numBlocksToFetch,
		)
	}
```

This should only happen if their `BLOCK_POLLING_INTERVAL` was manually set to a very infrequent interval.

Reply:
Or if they experience network issues or Infura/Alchemy/someone's own Ethereum node goes down temporarily. Shouldn't we just do the same thing here that we do for "fast sync"?

Reply:
Went with returning an error and having Mesh crash, since in addition to re-setting the blocks stored in the DB, we'd also need OrderWatcher to re-validate all orders at the latest block. This will be done upon Mesh re-boot.